mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/memory/DeleteByQueryExecutor.java

```diff
@@ -0,0 +1,88 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.vault.memory;
+
+import org.apache.james.core.User;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.task.Task;
+import org.apache.james.util.FunctionalUtils;
+import org.apache.james.vault.DeletedMessageVault;
+import org.apache.james.vault.search.Query;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class DeleteByQueryExecutor {
+    static class Notifiers {
+        private final Runnable userHandledNotifier;
+        private final Runnable searchErrorNotifier;
+        private final Runnable deletionErrorNotifier;
+        private final Runnable permanentlyDeletedMessageNotifyer;
+
+        Notifiers(Runnable userHandledNotifier, Runnable searchErrorNotifier, Runnable deletionErrorNotifier, Runnable permanentlyDeletedMessageNotifyer) {
+            this.userHandledNotifier = userHandledNotifier;
+            this.searchErrorNotifier = searchErrorNotifier;
+            this.deletionErrorNotifier = deletionErrorNotifier;
+            this.permanentlyDeletedMessageNotifyer = permanentlyDeletedMessageNotifyer;
+        }
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DeleteByQueryExecutor.class);
+
+    private final DeletedMessageVault deletedMessageVault;
+
+    public DeleteByQueryExecutor(DeletedMessageVault deletedMessageVault) {
+        this.deletedMessageVault = deletedMessageVault;
+    }
+
+    public Task.Result deleteByQuery(Query allBeforeRetentionPeriod, Notifiers notifiers) {
+        return Flux.from(deletedMessageVault.usersWithVault())
+            .flatMap(user -> deleteByQueryForUser(allBeforeRetentionPeriod, user, notifiers))
```

Comment:
I don't understand why the query has this name. If you accept a query, it can be anything, right?

Reply:
true, refactoring miss